Observer/BackfillTransactions.php

```diff
@@ -15,309 +15,321 @@
         \Magento\Framework\DataObject\IdentityGeneratorInterface $identityService,
         \Magento\Framework\Serialize\SerializerInterface $serializer,
-        \Magento\Authorization\Model\UserContextInterface $userContext
+        \Magento\Authorization\Model\UserContextInterface $userContext,
+        \Taxjar\SalesTax\Model\Configuration $taxjarConfig
     ) {
         $this->request = $request;
-        $this->storeManager = $storeManager;
-        $this->logger = $logger->setFilename(TaxjarConfig::TAXJAR_TRANSACTIONS_LOG)->force();
+        $this->logger = $logger->setFilename(\Taxjar\SalesTax\Model\Configuration::TAXJAR_TRANSACTIONS_LOG)->force();
         $this->orderRepository = $orderRepository;
-        $this->filterBuilder = $filterBuilder;
+        $this->storeManager = $storeManager;
         $this->searchCriteriaBuilder = $searchCriteriaBuilder;
-        $this->taxjarConfig = $taxjarConfig;
         $this->bulkManagement = $bulkManagement;
         $this->operationFactory = $operationFactory;
-        $this->identityService = $identityService;
         $this->serializer = $serializer;
         $this->userContext = $userContext;
+        $this->taxjarConfig = $taxjarConfig;
+        $this->uuid = $identityService->generateId();
     }
 
     /**
-     * @param Observer $observer
-     * @throws LocalizedException
+     * @throws \Magento\Framework\Exception\LocalizedException
      */
-    public function execute(Observer $observer): void
+    public function execute(\Magento\Framework\Event\Observer $observer)
     {
-        $this->apiKey = $this->taxjarConfig->getApiKey();
+        $this->observer = $observer;
 
-        if (!$this->apiKey) {
-            throw new LocalizedException(
-                __('Could not sync transactions with TaxJar. Please make sure you have an API key.')
-            );
+        try {
+            if (!$this->taxjarConfig->getApiKey()) {
+                throw new \Magento\Framework\Exception\LocalizedException(
+                    __('Could not sync transactions with TaxJar. Please make sure you have an API key.')
+                );
+            }
+
+            $criteria = $this->getSearchCriteria();
+            $orders = $this->getOrders($criteria, (bool)$this->getInput('force'));
+
+            if (!empty($orders)) {
+                $this->syncTransactions($orders);
+            }
+
+            $this->success();
+        } catch (\Exception $e) {
+            $this->fail($e);
         }
+    }
 
-        $this->logger->log(__('Initializing TaxJar transaction sync'));
+    /**
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function success()
+    {
+        $message = 'Transaction sync successfully scheduled. ';
+        $message .= json_encode($this->getConfiguration());
 
-        $criteria = $this->getSearchCriteria($observer->getData());
-        $orderResult = $this->orderRepository->getList($criteria);
+        $this->logger->log($message);
```

Comment:
Let's separate the log message from what get's displayed in the UI. In the UI I think the successful message should be something like:

"Successfully scheduled 5 transactions to be synced to TaxJar."

Then for the log we could included that message but then have the additional configuration details. Something like this perhaps:

"Successfully scheduled 5 transactions to be synced to TaxJar. Details: {"from":sometimestamp,"to":sometimestamp,"force_sync":false}"

In the failure case we should do them same. Give the failure message and the exemption message back to the UI. Then for the log entry give those messages plus the configuration data.

Reply:
I agree with the premise, but this one is easier said than done, because what is shown in the UI is the `playback()` of console output for our Logger class, kind of just "using" a side-effect of how `Logger::class` works.  To properly handle this situation, I would have change functionality of `Logger::class`, which might be out of scope.

You can see how the response we display in UI is generated here:
![image](https://user-images.githubusercontent.com/47947793/141869035-7fd22a34-9774-49b3-ac82-eb0a834ded0b.png)

If we wanted to continue using the Logger class, this could probably be accomplished pretty easily with either new method(s) on the Logger class, or refactor what is already there, but might also be worth more discussion since we are already planning other changes to logging.

Reply:
Resolved with addition of `playback` setter-method on Logger class and refactor of BackfillTransactions success and fail methods.